File: oop/04-abstraction/typescript/savingsAccount.ts
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: SavingsAccount implements all abstract methods from Account
 */

import { Account } from "./account";

export class SavingsAccount extends Account {
    public readonly interestRate: number;
    private static readonly MONTHLY_FEE = 0;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number, interestRate: number) {
        super(accountNumber, accountHolder, openingBalance);
        this.interestRate = interestRate;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Withdrawal amount must be greater than zero.");
        }
        if (amount > this._balance) {
            throw new Error(`Insufficient funds. Available: R${this._balance.toFixed(2)}`);
        }
        this._balance -= amount;
    }

    calculateMonthlyFees(): number {
        return SavingsAccount.MONTHLY_FEE;
    }

    getAccountSummary(): string {
        return `Savings Account ${this.accountNumber} | Holder: ${this.accountHolder} | Balance: R${this._balance.toFixed(2)} | Interest Rate: ${(this.interestRate * 100).toFixed(2)}%`;
    }

    applyInterest(): void {
        const interest = this._balance * this.interestRate;
        this._balance += interest;
    }
}

File: oop/04-abstraction/csharp/SavingsAccount.cs
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: SavingsAccount implements all abstract methods from Account
 */

namespace FinanceDomain.Abstraction;

public class SavingsAccount : Account
{
    public decimal InterestRate { get; }
    private const decimal MonthlyFee = 0m;

    public SavingsAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal interestRate)
        : base(accountNumber, accountHolder, openingBalance)
    {
        InterestRate = interestRate;
    }

    public override void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Withdrawal amount must be greater than zero.");

        if (amount > _balance)
            throw new InvalidOperationException($"Insufficient funds. Available: R{_balance:F2}");

        _balance -= amount;
    }

    public override decimal CalculateMonthlyFees()
    {
        return MonthlyFee;
    }

    public override string GetAccountSummary()
    {
        return $"Savings Account {AccountNumber} | Holder: {AccountHolder} | Balance: R{_balance:F2} | Interest Rate: {InterestRate:P2}";
    }

    public void ApplyInterest()
    {
        decimal interest = _balance * InterestRate;
        _balance += interest;
    }
}
